api/content.go

```diff
@@ -0,0 +1,202 @@
+package api
+
+import (
+	"net/http"
+	"sort"
+
+	dprequest "github.com/ONSdigital/dp-net/request"
+	"github.com/ONSdigital/dp-topic-api/apierrors"
+	"github.com/ONSdigital/dp-topic-api/models"
+	"github.com/ONSdigital/log.go/log"
+	"github.com/gorilla/mux"
+)
+
+func addItem(contentList *models.ContentResponseAPI, typeName string, itemLink *[]models.TypeLinkObject, id string, state string, privateResponse bool) {
+	var count int
+
+	if itemLink == nil {
+		return
+	}
+
+	nofItems := len(*itemLink)
+	if nofItems == 0 {
+		return
+	}
+
+	// Create list of sorted href's from itemLink list
+	hrefs := make([]string, 0, nofItems)
+	for _, field := range *itemLink {
+		hrefs = append(hrefs, field.HRef)
+	}
+	sort.Strings(hrefs)
+
+	// Iterate through sorted hrefs and use each one to select item from
+	// itemLink in alphabetical order
+	for _, href := range hrefs {
+		for _, item := range *itemLink {
+			if href == item.HRef {
+				var topicLink models.LinkObject
+				var selfLink models.LinkObject
+
+				selfLink.HRef = item.HRef
+				topicLink.ID = id
+				topicLink.HRef = "/topic/" + id
+
+				var cLinks models.ContentLinks
+				cLinks.Self = &selfLink
+				cLinks.Topic = &topicLink
+
+				var cItem models.ContentItem
+				cItem.Title = item.Title
+				cItem.Type = typeName
+				if privateResponse {
+					cItem.State = state
+				}
+				cItem.Links = &cLinks
+
+				if contentList.Items == nil {
+					contentList.Items = &[]models.ContentItem{cItem}
+				} else {
+					*contentList.Items = append(*contentList.Items, cItem)
+				}
+
+				count++
+			}
+		}
+	}
+
+	contentList.TotalCount = contentList.TotalCount + count
+}
+
+// getContentPublicHandler is a handler that gets content by its id from MongoDB for Web
+func (api *API) getContentPublicHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	vars := mux.Vars(req)
+	id := vars["id"]
+	logdata := log.Data{
+		"request_id": ctx.Value(dprequest.RequestIdKey),
+		"content_id": id,
+		"function":   "getContentPublicHandler",
+	}
+
+	// check topic from mongoDB by id
+	_, err := api.dataStore.Backend.GetTopic(id)
+	if err != nil {
+		handleError(ctx, w, err, logdata)
+		return
+	}
+
+	// get content from mongoDB by id
+	content, err := api.dataStore.Backend.GetContent(id)
+	if err != nil {
+		// no content found
+		handleError(ctx, w, err, logdata)
+		return
+	}
+
+	// User is not authenticated and hence has only access to current sub document(s)
+
+	if content.Current == nil {
+		handleError(ctx, w, apierrors.ErrInternalServer, logdata)
```

Comment:
In public handlers, if there is no current document we should just return a not found. Having a different response to other non-existant resources points out that there might be something internally at that address, but since it hasnt been made public yet its a security issue to give users that knowledge

Reply:
Umm... i copied this logic / error message from what we previously decided upon in the function:
getSubtopicsPublicHandler()
in api/topic.go

Reply:
It should be changed in all places. The lack of a current doc can never be anything but a 404 to public users on any handler

Reply:
After chat, code changed to use ErrContentNotFound